### ISSUE TITLE
RFC: moving json column to jsonb

### ISSUE THREAD

**Issue by jmartinm** [inveniosoftware/invenio-records#171]
Proposal
======

Move `RecordMetadata` json column to `JSONB` when using `postgresql`.

The following code:

```
json = db.Column(
        JSONType().with_variant(
            postgresql.JSON(none_as_null=True),
            'postgresql',
        ),
        default=lambda: dict(),
        nullable=True
    )
```

Can be converted to

```

json = db.Column(
        db.JSON().with_variant(
            postgresql.JSONB(none_as_null=True),
            'postgresql',
        ),
        default=lambda: dict(),
        nullable=True
    )
```

This will allow to use native `jsonb` column in PostgreSQL (see https://www.postgresql.org/docs/9.4/static/datatype-json.html) with advantages such as better indexing capabilities and using containment operators.

-----------

I have tested locally with `PostgreSQL v9.5` and:

```
Flask-SQLAlchemy==2.2
SQLAlchemy==1.1.10
SQLAlchemy-Continuum==1.3.0
SQLAlchemy-Utils==0.32.14
```

And everything seems to work as expected, enabling queries such as:

```
db.session.query(RecordMetadata).filter(type_coerce(RecordMetadata.json, JSONB()).contains({"control_number": 1498185})).all()
```

(Note that `type_coerce` is needed because of the `with_variant` construct - [more info](https://bitbucket.org/zzzeek/sqlalchemy/issues/3855/comparator-on-variant-field)

Which translates to:

```
SELECT *
FROM records_metadata
WHERE records_metadata.json @> '{"control_number": 1498185}'
```

----------------------------

The only thing missing to be done is to write an Alembic recipe to migrate the column.

cc @lnielsen 

Related to https://github.com/inveniosoftware/invenio-db/issues/19

**Comment by slint** [comment]
Different dialects will require different upgrade recipes though:

* **Postgres**: Simple as `op.alter_column('record_metadat', 'json', type_=postgresql.JSONB, postgresql_using='json::text::jsonb')`
* **MySQL**: [*A bit more elaborate actually...*](https://stackoverflow.com/a/37820319)

**Comment by jmartinm** [comment]
> MySQL: A bit more elaborate actually...

Since we are only altering the case of `postgresql` to move from `json` to `jsonb` do we need a migration for the MySQL dialect?

**Comment by slint** [comment]
> Since we are only altering the case of postgresql to move from json to jsonb do we need a migration for the MySQL dialect?

Might be needed, since previously `sqlalchemy_utils.types.JSONType` was used which [defaulted to `sa.UnicodeText`](http://sqlalchemy-utils.readthedocs.io/en/latest/_modules/sqlalchemy_utils/types/json.html), and now [`sa.JSON`](http://docs.sqlalchemy.org/en/latest/core/type_basics.html#sqlalchemy.types.JSON) is used.

**Comment by kaplun** [comment]
Mmh. Is there any one willing to use Invenio 3 on MySQL? All services on 3 are on PostgreSQL. If supporting MySQL means complicating support for PostgreSQL, maybe we could leave it behind?

**Comment by lnielsen** [comment]
@kaplun Are you talking about just the upgrade recipe for MySQL or in general about MySQL support in Invenio 3?

If it's just the upgrade receipe, then I'm fine with that it's only for PostgreSQL as we don't have stable releases out yet. 

In general for Invenio v3 though, we should support both MySQL and PostgreSQL (and SQLite for testing). There is not much more work in supporting both, and it ensures that we avoid lock-in on the database level like in Invenio v1. Invenio v3 is likely going to have a life-time of 15-20 years so many things can happen during this time period.

**Comment by jmartinm** [comment]
See WIP https://github.com/inveniosoftware/invenio-records/pull/172

**Comment by drjova** [comment]
Closed by 342c8def93e17206a0cd17feb99d119d85b6b74a
